hub/src/server/server.ts

```diff
@@ -74,6 +75,23 @@ export class HubServer {
     }
   }
 
+  getHistoricalFileName(address: string, filePath: string) {
+    const pathParts = filePath.split('/')
+    const fileName = pathParts[pathParts.length - 1]
+    const filePathPrefix = filePath.slice(0, filePath.length - fileName.length)
+  
+    // reject writes to historical files
+    if (fileName.startsWith('.history.')) {
```

Comment:
I'm not sure why this check is here.  The file's name should be able to be anything, since we're going to prefix it with `.history.${Date.now()}.${generateUniqueID()}.`, right?

Reply:
There was a concern at one point to prevent restricted apps from deleting or zeroing out historical files, but that should not be possible with current implementation. Attempting to do so would just create a historical file for a historical file.

So, this check just prevents that historical-file-inception. I think this is would be more an accidental thing an app dev might do. For example listFiles, iterate, write to a historical file, say `.history.1564526417230.nTXJqe7gTc.myphoto.png`, and trigger a historical rename of the historical file: `history.1564526772589.C9H8GzbAlA..history.1564526417230.nTXJqe7gTc.myphoto.png`

But doesn't seem like that big of a deal. Should we allow it?

Reply:
This also brings up the question -- should the existence of a `putFileArchival` scope also restrict `list-files`? For example, not returning entries for any of the historical files. 

One oddity would be that a user's bucket may contain >PAGE_LIMIT historical files, and end up with lots of `list-file` responses that contain a pagination marker but no file entries.

Reply:
> So, this check just prevents that historical-file-inception.

I don't think this is a concern.  A buggy app is a buggy app.

> This also brings up the question -- should the existence of a putFileArchival scope also restrict list-files? For example, not returning entries for any of the historical files.

At the very least, I think `list-files` should hide the timestamp unless it had `putFileArchival` permission, since this would allow an app to view the user's activity.  But I take your point about there being an inexplicably large number of empty pages.  Maybe `list-files` should censor the name and timestamp of history files so they all say `<hidden>` and have a timestamp of `0`?

Reply:
@yknl Any thoughts on this?

Reply:
I think on these two questions,

1) We should allow historical naming inception. It's not a safety concern, so why add this caveat to the data model (i.e., you can name your file anything but something starting with `.history.`).
2) Historical files _should_ appear in `list-files` for the time being. `list-files` isn't really usable on the typical read path for applications (because it doesn't really work for multi-player reads), but it's primary use is for supporting (a) backups, and, in the case of archival puts, (b) rollbacks. So, if an application wanted to actually find an archived version of a file, it could do so using `list-files`. If the files were hidden, that wouldn't be possible.

Reply:
The `.history.` name prefix restriction has been removed.

Reply:
My understanding is that list-file with an archival scoped auth token should not return historical files. When the file manager in the browser is implemented, it can use a full scope auth token to view historical files.

Reply:
In the collections meeting we decided to not include historical files in list-files with an archival scope -- PR has been updated.

Reply:
The blockstack.js `listFiles` loop will break with empty pages due to historical file filtering https://github.com/blockstack/blockstack.js/blob/76a1fb23ffc7db6ebb371698057da8baa48e8434/src/storage/index.ts#L696

Not sure how to fix this cleanly without also updating blockstack.js. Gaia could detect this situation and return a single 'null' item in the entries array which would cause the listFile loop to continue, but probably break apps that are not expecting null entries. Thoughts?

Reply:
I continue to think that the historical files should show up in list-files. What's the downside to this? Do applications use list-files for any functionality?

Reply:
I think it's mainly a privacy concern -- an application can see the historic states of a collection file.

Reply:
@kantai Yeah some privacy concerns.

Yes apps do use listFiles:
https://github.com/search?utf8=%E2%9C%93&q=%22blockstack%22+%22listfiles%22&type=Code&ref=advsearch&l=&l=

Apps or blockstack.js would need to implement parsing & handling of the historical file naming scheme. This will probably need to be done at least in the Browser at some point anyway for the restoration tool, but that can be done later.